app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt

```diff
@@ -66,48 +81,89 @@ class JupiterSwapPresenter(
     }
 
     private fun handleLoadingTransaction(state: SwapState.LoadingTransaction) {
-        view?.setFirstTokenWidgetState(
-            state = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = state.amountTokenA)
-        )
-        view?.setSecondTokenWidgetState(
-            state = widgetMapper.mapTokenB(token = state.tokenB, tokenAmount = state.amountTokenB)
-        )
+        widgetAState = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = state.amountTokenA)
+        widgetBState = widgetMapper.mapTokenB(token = state.tokenB, tokenAmount = state.amountTokenB)
+        updateWidgets()
         view?.setButtonState(buttonState = buttonMapper.mapLoading())
+        getRateTokenA(tokenA = state.tokenA, tokenAmount = state.amountTokenA)
+        getRateTokenB(tokenB = state.tokenB, tokenAmount = state.amountTokenB)
     }
 
     private fun handleLoadingRoutes(state: SwapState.LoadingRoutes) {
-        view?.setFirstTokenWidgetState(
-            state = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = state.amountTokenA)
-        )
-        view?.setSecondTokenWidgetState(state = widgetMapper.mapTokenBLoading(token = state.tokenB))
+        widgetAState = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = state.amountTokenA)
+        widgetBState = widgetMapper.mapTokenBLoading(token = state.tokenB)
+        updateWidgets()
         view?.setButtonState(buttonState = buttonMapper.mapLoading())
+        getRateTokenA(tokenA = state.tokenA, tokenAmount = state.amountTokenA)
     }
 
     private fun handleTokenAZero(state: SwapState.TokenAZero) {
-        view?.setFirstTokenWidgetState(
-            state = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = null)
-        )
-        view?.setSecondTokenWidgetState(
-            state = widgetMapper.mapTokenB(token = state.tokenB, tokenAmount = null)
-        )
+        widgetAState = widgetMapper.mapTokenA(token = state.tokenA, tokenAmount = null)
+        widgetBState = widgetMapper.mapTokenB(token = state.tokenB, tokenAmount = null)
+        updateWidgets()
         view?.setButtonState(buttonMapper.mapEnterAmount())
     }
 
     private fun handleInitialLoading() {
-        view?.setFirstTokenWidgetState(state = widgetMapper.mapWidgetLoading(isTokenA = true))
-        view?.setSecondTokenWidgetState(state = widgetMapper.mapWidgetLoading(isTokenA = false))
+        widgetAState = widgetMapper.mapWidgetLoading(isTokenA = true)
+        widgetBState = widgetMapper.mapWidgetLoading(isTokenA = false)
+        updateWidgets()
         view?.setButtonState(buttonState = SwapButtonState.Hide)
     }
+
+    private fun getRateTokenA(tokenA: SwapTokenModel, tokenAmount: BigDecimal) {
+        rateTokenAJob?.cancel()
+        rateTokenAJob = rateLoaderTokenA.getRate(tokenA)
+            .flowOn(dispatchers.io)
+            .onEach { handleRateLoader(state = it, isTokenA = true, tokenAmount = tokenAmount) }
```

Comment:
maybe you can pass some enum value and remove boolean condition here and in mapWidgetLoading?